readme.md
# postcss-px2vp

[![npm version](https://badge.fury.io/js/postcss-px2vp.svg)](https://badge.fury.io/js/postcss-px2vp)

将 px 单位转换为视口单位的 (vw, vh, vmin, vmax) 的 [PostCSS](https://github.com/postcss/postcss) 插件.

## 简介

如果你的样式需要做根据视口大小来调整宽度，这个脚本可以将你 CSS 中的 px 单位转化为 vw，1vw 等于 1/100 视口宽度。

### 输入

```css
.class {
  margin: -10px 0.5vh;
  padding: 5vmin 9.5px 1px;
  border: 3px solid black;
  border-bottom-width: 1px;
  font-size: 14px;
  line-height: 20px;
}
.class2 {
  border: 1px solid black;
  margin-bottom: 1px;
  font-size: 20px; /* px-to-viewport-ignore */
  /* px-to-viewport-ignore-next */
  line-height: 30px;
}
@media (min-width: 750px) {
  .class3 {
    font-size: 16px;
    line-height: 22px;
  }
}
```

### 输出

```css
.class {
  margin: -1.33333vw 0.5vh;
  padding: 5vmin 1.26667vw 1px;
  border: 0.4vw solid black;
  border-bottom-width: 1px;
  font-size: 1.86667vw;
  line-height: 2.66667vw;
}
.class2 {
  border: 1px solid black;
  margin-bottom: 1px;
  font-size: 20px;
  line-height: 30px;
}
@media (min-width: 750px) {
  .class3 {
    font-size: 2.13333vw;
    line-height: 2.93333vw;
  }
}

```

## 上手

### 安装

使用 npm 安装

```
npm install postcss-px2vp --save-dev
```

或者使用 yarn 进行安装

```
yarn add -D postcss-px2vp
```

### 配置参数

默认参数:

```js
{
  unitToConvert: 'px',
  viewportWidth: 320,
  unitPrecision: 5,
  propList: ['*'],
  viewportUnit: 'vw',
  fontViewportUnit: 'vw',
  selectorBlackList: [],
  minPixelValue: 1,
  mediaQuery: false,
  replace: true,
  exclude: [],
  landscape: false,
  landscapeUnit: 'vw',
  landscapeWidth: 568
}
```

- `unitToConvert` (String) 需要转换的单位，默认为"px"
- `viewportWidth` (Number) 设计稿的视口宽度
- `unitPrecision` (Number) 单位转换后保留的精度
- `propList` (Array) 能转化为 vw 的属性列表
  - 传入特定的 CSS 属性；
  - 可以传入通配符"_"去匹配所有属性，例如：['_']；
  - 在属性的前或后添加"*",可以匹配特定的属性. (例如['*position\*'] 会匹配 background-position-y)
  - 在特定属性前加 "!"，将不转换该属性的单位 . 例如: ['*', '!letter-spacing']，将不转换 letter-spacing
  - "!" 和 "_"可以组合使用， 例如: ['_', '!font\*']，将不转换 font-size 以及 font-weight 等属性
- `viewportUnit` (String) 希望使用的视口单位
- `fontViewportUnit` (String) 字体使用的视口单位
- `selectorBlackList` (Array) 需要忽略的 CSS 选择器，不会转为视口单位，使用原有的 px 等单位。
  - 如果传入的值为字符串的话，只要选择器中含有传入值就会被匹配
    - 例如 `selectorBlackList` 为 `['body']` 的话， 那么 `.body-class` 就会被忽略
  - 如果传入的值为正则表达式的话，那么就会依据 CSS 选择器是否匹配该正则
    - 例如 `selectorBlackList` 为 `[/^body$/]` , 那么 `body` 会被忽略，而 `.body` 不会
- `minPixelValue` (Number) 设置最小的转换数值，如果为 1 的话，只有大于 1 的值会被转换
- `mediaQuery` (Boolean) 媒体查询里的单位是否需要转换单位
- `replace` (Boolean) 是否直接更换属性值，而不添加备用属性
- `exclude` (Array or Regexp) 忽略某些文件夹下的文件或特定文件，例如 'node_modules' 下的文件
  - 如果值是一个正则表达式，那么匹配这个正则的文件会被忽略
  - 如果传入的值是一个数组，那么数组里的值必须为正则
- `landscape` (Boolean) 是否添加根据 `landscapeWidth` 生成的媒体查询条件 `@media (orientation: landscape)`
- `landscapeUnit` (String) 横屏时使用的单位
- `landscapeWidth` (Number) 横屏时使用的视口宽度  
  **P.S. 所有参数都可以传入一个函数，动态改变参数**
  示例

  ```typescript
    {
      viewportWidth(rule: PostCss.Rule){
        const file = rule.source?.input.file;
        if (file?.includes('main')) return 750;
        return 375;
      }
    }
  ```

#### 直接在 gulp 中使用，添加 gulp-postcss

在 `gulpfile.js` 添加如下配置:

```js
var gulp = require('gulp');
var postcss = require('gulp-postcss');
var pxtoviewport = require('postcss-px2vp');

gulp.task('css', function () {
  var processors = [
    pxtoviewport({
      viewportWidth: 320,
      viewportUnit: 'vmin'
    })
  ];

  return gulp
    .src(['build/css/**/*.css'])
    .pipe(postcss(processors))
    .pipe(gulp.dest('build/css'));
});
```

#### 使用 PostCss 配置文件时

在`postcss.config.js`添加如下配置

```js
module.exports = {
  plugins: {
    ...
    'postcss-px2vp': {
      // options
    }
  }
}
```

## Changelog

变更日志在 [这](CHANGELOG.md).

## 许可

本项目使用 [MIT License](LICENSE).

## 借鉴至

本项目基本逻辑都是从[postcss-px-to-viewport](https://github.com/evrone/postcss-px-to-viewport/)的 clone 过来。
本项目主要做了如下工作：

- 以 typescript 重构源码；
- 添加参数动态化配置方法；
- 以 Postcss 8 推荐的方法重写了插件声明形式。


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.3](https://github.com/sexyHuang/postcss-px2vp/compare/v1.1.2...v1.1.3) (2022-03-08)


### Bug Fixes

* **#3:** 修复ignore comment 不生效问题 ([4ec1e30](https://github.com/sexyHuang/postcss-px2vp/commit/4ec1e301b8a6f1d88bfc281920ac6b4c78e77c18)), closes [#3](https://github.com/sexyHuang/postcss-px2vp/issues/3)

### [1.1.2](https://github.com/sexyHuang/postcss-px2vp/compare/v1.1.1...v1.1.2) (2021-05-13)

### [1.1.1](https://github.com/sexyHuang/postcss-px2vp/compare/v1.1.0...v1.1.1) (2021-05-13)

## [1.1.0](https://github.com/sexyHuang/postcss-px2vp/compare/v1.1.0-beta.0...v1.1.0) (2021-05-13)

## [1.1.0-beta.0](https://github.com/sexyHuang/postcss-px2vp/compare/v1.1.0-alpha.1...v1.1.0-beta.0) (2021-05-13)

## [1.1.0-alpha.1](https://github.com/sexyHuang/postcss-px2vp/compare/v1.1.0-alpha.0...v1.1.0-alpha.1) (2021-05-13)


### Features

* 修改返回类型 ([59a96db](https://github.com/sexyHuang/postcss-px2vp/commit/59a96dbfdc900da7ada99a19a752925676832d17))

## 1.1.0-alpha.0 (2021-05-13)


### Features

* 功能完成 ([fd03d99](https://github.com/sexyHuang/postcss-px2vp/commit/fd03d99802caf510799365658cab71ce4784f2a2))
* tools ([c4c1178](https://github.com/sexyHuang/postcss-px2vp/commit/c4c1178f2b1f2a63ab1fc6b0209a78076fc47f07))


src/index.ts
import type { Container, Declaration, Plugin, Rule } from 'postcss';
import getUnitRegexp from './pixel-unit-regexp';
import { createPropListMatcher } from './prop-list-matcher';

type ViewportUnit = 'vw' | 'vh' | 'vmin' | 'vmax';

type Option = {
  /** 需要转换的单位，默认为"px" */
  unitToConvert?: string;
  /** 设计稿的视口宽度，默认为320 */
  viewportWidth?: number;

  viewportHeight?: number; // not now used; TODO: need for different units and math for different properties
  /** 单位转换后保留的精度，默认5*/
  unitPrecision?: number;
  /** 希望使用的视口单位，默认为"vw" */
  viewportUnit?: ViewportUnit;
  /** 字体使用的视口单位，默认为"vw" */
  fontViewportUnit?: ViewportUnit; // vmin is more suitable.
  /** 需要忽略的CSS选择器，不会转为视口单位，使用原有的px等单位。 */
  selectorBlackList?: (string | RegExp)[];
  /**
   * 能转化为vw的属性列表
   * - 传入特定的CSS属性；
   * - 可以传入通配符"*"去匹配所有属性，例如：['*']；
   * - 在属性的前或后添加"*",可以匹配特定的属性. (例如['*position*'] 会匹配 background-position-y)
   * - 在特定属性前加 "!"，将不转换该属性的单位 . 例如: ['*', '!letter-spacing']，将不转换letter-spacing
   * - "!" 和 "*"可以组合使用， 例如: ['*', '!font*']，将不转换font-size以及font-weight等属性
   */
  propList?: string[];
  /** 设置最小的转换数值，默认为1 */
  minPixelValue?: number;
  /** 媒体查询里的单位是否需要转换单位，默认false */
  mediaQuery?: boolean;
  /** 是否直接更换属性值，而不添加备用属性，默认true */
  replace?: boolean;
  /** 是否添加根据 `landscapeWidth` 生成的媒体查询条件 `@media (orientation: landscape)`，默认false */
  landscape?: boolean;
  /** 横屏时使用的单位，默认"vw" */
  landscapeUnit?: ViewportUnit;
  /** 忽略某些文件夹下的文件或特定文件 */
  exclude?: RegExp | RegExp[];
  /** 横屏时使用的视口宽度，默认568 */
  landscapeWidth?: number;
};

const getDefault = () =>
  ({
    unitToConvert: 'px',
    viewportWidth: 320,
    viewportHeight: 568, // not now used; TODO: need for different units and math for different properties
    unitPrecision: 5,
    viewportUnit: 'vw',
    fontViewportUnit: 'vw', // vmin is more suitable.
    selectorBlackList: [],
    propList: ['*'],
    minPixelValue: 1,
    mediaQuery: false,
    replace: true,
    landscape: false,
    landscapeUnit: 'vw',
    exclude: [],
    landscapeWidth: 568
  } as Required<Option>);

const postcssPlugin = 'postcss-px2vp';

const IGNORE_NEXT_COMMENT = 'px-to-viewport-ignore-next';
const IGNORE_PREV_COMMENT = 'px-to-viewport-ignore';

type FunctionalOptions<T> = {
  [K in keyof T]: T[K] | ((rule: Rule) => T[K]);
};

type Input = FunctionalOptions<Option>;

const isExclude = (exclude: RegExp | RegExp[], file?: string) => {
  if (!file) return false;
  if (Array.isArray(exclude)) return exclude.some(reg => reg.test(file));
  return exclude.test(file);
};

function getOption<T>(
  option: T,
  rule: Rule
): T extends (rule: Rule) => infer K ? K : T {
  return typeof option === 'function' ? option(rule) : option;
}

function toFixed(number: number, precision: number) {
  var multiplier = Math.pow(10, precision + 1),
    wholeNumber = Math.floor(number * multiplier);
  return (Math.round(wholeNumber / 10) * 10) / multiplier;
}

function createPxReplace(
  opts: { minPixelValue: number; unitPrecision: number },
  viewportUnit: ViewportUnit,
  viewportSize: number
) {
  return function (m: string, $1: string) {
    if (!$1) return m;
    var pixels = parseFloat($1);
    if (pixels <= opts.minPixelValue) return m;
    var parsedVal = toFixed((pixels / viewportSize) * 100, opts.unitPrecision);
    return parsedVal === 0 ? '0' : `${parsedVal}${viewportUnit}`;
  };
}

function validateParams(params: undefined | string, mediaQuery: boolean) {
  return !params || (params && mediaQuery);
}

function getUnit(
  prop: string,
  opts: { viewportUnit: ViewportUnit; fontViewportUnit: ViewportUnit }
) {
  return !prop.includes('font') ? opts.viewportUnit : opts.fontViewportUnit;
}

function declarationExists(
  decls: Container | undefined,
  prop: string,
  value: string
) {
  if (!decls) return false;
  return decls.some(function (decl: Declaration) {
    return decl.prop === prop && decl.value === value;
  });
}

function optionCreator({
  options,
  rule,
  defaultOptions
}: {
  options?: Input;
  rule: Rule;
  defaultOptions: Required<Option>;
}): Required<Option> {
  if (!options) return defaultOptions;
  return {
    ...defaultOptions,
    ...Object.entries(options).reduce((prev, [key, value]) => {
      prev[key] = getOption(value, rule);
      return prev;
    }, {} as any)
  };
}

const blacklistedSelector = (
  blacklist: (string | RegExp)[],
  selector: string
) =>
  blacklist.some(rule => {
    if (typeof rule === 'string') return selector.includes(rule);
    return rule.test(selector);
  });

interface PluginCreator<PluginOptions> {
  (options?: PluginOptions): Plugin;
  postcss: true;
}

const px2vp: PluginCreator<Input> = options => {
  const landscapeRules: Rule[] = [];
  const defaultOptions = getDefault();

  return {
    postcssPlugin,
    Once(root, { atRule, result }) {
      root.walkRules(rule => {
        const file = rule.source?.input.file;

        const {
          exclude,
          selectorBlackList,
          propList,
          landscape,
          unitToConvert,
          minPixelValue,
          unitPrecision,
          landscapeUnit,
          landscapeWidth,
          fontViewportUnit,
          viewportUnit,
          mediaQuery,
          viewportWidth,
          replace
        } = optionCreator({ options, rule, defaultOptions });
        // init options
        const pxRegex = getUnitRegexp(unitToConvert);
        const satisfyPropList = createPropListMatcher(propList);
        const params = (rule.parent as any)?.params as string | undefined;
        if (
          isExclude(exclude, file) ||
          blacklistedSelector(selectorBlackList, rule.selector)
        )
          return;

        if (landscape && !params) {
          const landscapeRule = rule.clone().removeAll();
          rule.walkDecls(decl => {
            const { value, prop } = decl;
            if (!value.includes(unitToConvert) || !satisfyPropList(prop))
              return;

            landscapeRule.append(
              decl.clone({
                value: value.replace(
                  pxRegex,
                  createPxReplace(
                    { minPixelValue, unitPrecision },
                    landscapeUnit,
                    landscapeWidth
                  )
                )
              })
            );
            if (landscapeRule.nodes.length > 0) {
              landscapeRules.push(landscapeRule);
            }
          });
        }
        if (!validateParams(params, mediaQuery)) return;
        rule.walkDecls((decl, i) => {
          let { value, prop } = decl;
          if (!value.includes(unitToConvert) || !satisfyPropList(prop)) return;

          const prev = decl.prev();
          if (prev?.type === 'comment' && prev.text === IGNORE_NEXT_COMMENT) {
            prev.remove();
            return;
          }
          const next = decl.next();
          if (next?.type === 'comment' && next.text === IGNORE_PREV_COMMENT) {
            if (/\n/.test(next.raws.before ?? '')) {
              result.warn(
                'Unexpected comment /* ' +
                  IGNORE_PREV_COMMENT +
                  ' */ must be after declaration at same line.',
                { node: next }
              );
            } else {
              // remove comment
              next.remove();
              return;
            }
          }
          const [unit, size] =
            landscape && params?.includes('landscape')
              ? [landscapeUnit, landscapeWidth]
              : [
                  getUnit(prop, { viewportUnit, fontViewportUnit }),
                  viewportWidth
                ];
          value = value.replace(
            pxRegex,
            createPxReplace({ minPixelValue, unitPrecision }, unit, size)
          );
          if (declarationExists(decl.parent, prop, value)) return;
          if (replace) decl.value = value;
          else decl.parent?.insertAfter(i, decl.clone({ value: value }));
        });
      });
      if (landscapeRules.length) {
        const landscapeRoot = atRule({
          params: '(orientation: landscape)',
          name: 'media'
        });
        landscapeRules.forEach(rule => {
          landscapeRoot.append(rule);
        });
        root.append(landscapeRoot);
      }
    }
  };
};
px2vp.postcss = true;

export = px2vp;
